lib/src/annotate.rs

```diff
@@ -0,0 +1,374 @@
+                    file_name,
+                    file_cache,
+                    current_commit,
+                    &parent_commit,
+                )?;
+
+                let parent_commit_line_map = commit_line_map
+                    .entry(parent_commit.id().clone())
+                    .or_default();
+
+                for (current_line_number, parent_line_number) in same_line_map {
+                    if let Some(original_line_number) =
+                        current_commit_line_map.remove(&current_line_number)
+                    {
+                        // forward current line to parent commit since it is in common
+                        parent_commit_line_map.insert(parent_line_number, original_line_number);
+                    }
+                }
+                if parent_commit_line_map.is_empty() {
+                    commit_line_map.remove(parent_commit.id());
+                }
+            }
+        }
+        if !current_commit_line_map.is_empty() {
+            mark_lines_from_original(
+                original_line_map,
+                current_commit.id(),
+                current_commit_line_map,
+            );
+        }
+        let _ = file_cache.remove(current_commit.id());
+    }
+
+    Ok(())
+}
+
+/// For two versions of the same file, for all the lines in common, overwrite
+/// the new mapping in the results for the new commit. Let's say I have
+/// a file in commit A and commit B. We know that according to local_line_map,
+/// in commit A, line 3 corresponds to line 7 of the original file. Now, line 3
+/// in Commit A corresponds to line 6 in commit B. Then, we update
+/// local_line_map to say that "Commit B line 6 goes to line 7 of the original
+/// file". We repeat this for all lines in common in the two commits. For 2
+/// identical files, we bulk replace all mappings from commit A to commit B in
+/// local_line_map
+fn process_files_in_commits(
+    store: &Store,
+    file_name: &RepoPath,
+    file_cache: &mut FileCache,
+    current_commit: &Commit,
+    parent_commit: &Commit,
+) -> Result<HashMap<usize, usize>, BackendError> {
+    load_file_into_cache(
+        file_cache,
+        store,
+        current_commit.id(),
+        file_name,
+        &current_commit.tree()?,
+    )?;
+    load_file_into_cache(
+        file_cache,
+        store,
+        parent_commit.id(),
+        file_name,
+        &parent_commit.tree()?,
+    )?;
+
+    let current_contents = file_cache.get(current_commit.id()).unwrap();
+    let parent_contents = file_cache.get(parent_commit.id()).unwrap();
+
+    Ok(get_same_line_map(current_contents, parent_contents))
+}
+
+/// For two files, get a map of all lines in common (e.g. line 8 maps to line 9)
+fn get_same_line_map(current_contents: &[u8], parent_contents: &[u8]) -> HashMap<usize, usize> {
+    let mut result_map = HashMap::new();
+    let inputs = vec![current_contents, parent_contents];
+    let diff = Diff::by_line(&inputs);
```

Comment:
nit: `Diff::by_line([current_contents, parent_contents])` (Vec is no longer needed)